pkg/apis/provisioning/v1alpha1/provisioner.go

```diff
@@ -83,6 +83,10 @@ type Constraints struct {
 	// OperatingSystem constrains the underlying node operating system
 	// +optional
 	OperatingSystem *string `json:"operatingSystem,omitempty"`
+	// CapacityType constrains what type of capacity is launched by the Provisioner
+	// If unspecified, it will launch on-demand nodes.
+	// +optional
+	CapacityType *string `json:"capacityType,omitempty"`
```

Comment:
Is there an upstream kubernetes concept for this? If not, we probably want to bake this into the cloud provider instead

Reply:
There is not currently... but shouldn't there be?!  All the major cloud providers have this concept, so it feels like it should be top-level and a k8s thing.  

Looks like there was a very brief discussion on adding one: https://github.com/kubernetes/kubernetes/issues/35166 

If there were a "node.kubernetes.io/capacity-type: spot" , it would be possible to add the percentage to different capacity types in the deployment controller.

Reply:
Based on conversations today and reading Jacob's proposal in https://github.com/awslabs/karpenter/pull/330, I attempted a cloud provider implementation for `capacity-type`.